x-pack/packages/ml/data_utils/src/get_agg_intervals.ts

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { get } from 'lodash';
+
+import type { Client } from '@elastic/elasticsearch';
+import * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+
+import { KBN_FIELD_TYPES } from '@kbn/field-types';
+import { isPopulatedObject } from '@kbn/ml-is-populated-object';
+
+import { buildSamplerAggregation } from './build_sampler_aggregation';
+import { getSamplerAggregationsResponsePath } from './get_sampler_aggregations_response_path';
+import { stringHash } from './string_hash';
+
+// TODO Temporary type definition until we can import from `@kbn/core`.
+/**
+ * Client used to query the elasticsearch cluster.
+ *
+ * @public
+ */
+export type ElasticsearchClient = Omit<
+  Client,
+  'connectionPool' | 'serializer' | 'extend' | 'close' | 'diagnostic'
+>;
+
+const MAX_CHART_COLUMNS = 20;
+
+export interface HistogramField {
+  fieldName: string;
+  type: string;
+}
+
+interface NumericColumnStats {
+  interval: number;
+  min: number;
+  max: number;
+}
+type NumericColumnStatsMap = Record<string, NumericColumnStats>;
+
+export const getAggIntervals = async (
+  client: ElasticsearchClient,
+  indexPattern: string,
+  query: any,
```

Comment:
Will be safer to replace this `query: any` with QueryDslQueryContainer from elasticsearch type package